lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -925,52 +1001,53 @@ public function andHaving($predicate, ...$predicates) : self
      */
     public function orHaving($predicate, ...$predicates) : self
     {
-        return $this->appendPredicates('having', CompositeExpression::TYPE_OR, $predicate, ...$predicates);
+        $this->having = $this->appendToPredicate($this->having, CompositeExpression::TYPE_OR, $predicate, ...$predicates);
+
+        $this->state = self::STATE_DIRTY;
+
+        return $this;
     }
 
     /**
-     * Sets one or more predicates combined by the AND logic as the given query clause.
-     * Replaces any previously specified predicates.
+     * Creates a CompositeExpression from one or more predicates combined by the AND logic.
      *
+     * @param string|CompositeExpression $predicate
      * @param string|CompositeExpression ...$predicates
      *
-     * @return $this This QueryBuilder instance.
+     * @return string|CompositeExpression
      */
-    private function setPredicates(string $clause, ...$predicates) : self
+    private function createPredicate($predicate, ...$predicates)
     {
-        if (count($predicates) > 1) {
-            $predicate = new CompositeExpression(
-                CompositeExpression::TYPE_AND,
-                $predicates
-            );
-        } else {
-            $predicate = array_shift($predicates);
+        if (! $predicates) {
+            return $predicate;
         }
 
-        return $this->add($clause, $predicate);
+        return new CompositeExpression(CompositeExpression::TYPE_AND, array_merge([$predicate], $predicates));
     }
 
     /**
-     * Appends the given predicates combined by the given type of logic to the given query clause.
+     * Appends the given predicates combined by the given type of logic to the current predicate.
      *
-     * @param string|CompositeExpression ...$predicates
+     * @param string|CompositeExpression|null $currentPredicate
+     * @param string|CompositeExpression      ...$predicates
      *
-     * @return $this This QueryBuilder instance.
+     * @return string|CompositeExpression
      */
-    private function appendPredicates(string $clause, string $type, ...$predicates) : self
+    private function appendToPredicate($currentPredicate, string $type, ...$predicates)
     {
-        $predicate = $this->getQueryPart($clause);
-
-        if ($predicate instanceof CompositeExpression && $predicate->getType() === $type) {
-            $predicate->addMultiple($predicates);
-        } else {
-            $predicate = new CompositeExpression(
-                $type,
-                array_merge([$predicate], $predicates)
-            );
+        if ($currentPredicate === null && count($predicates) === 1) {
+            return $predicates[0];
+        }
+
+        if ($currentPredicate instanceof CompositeExpression && $currentPredicate->getType() === $type) {
+            return $currentPredicate->addMultiple($predicates);
         }
 
-        return $this->add($clause, $predicate, true);
+        if ($currentPredicate !== null) {
```

Comment:
There are two inverse `if` statements: `if ($currentPredicate === null)` and `if ($currentPredicate !== null)`. Would it be possible to replace them with an `if`/`else`?

Reply:
Hardly. There's this `if (...) return` in-between, and if you move the `array_unshift()` above to merge into a `if...else`, you may end up with work being done for nothing if the function returns early.

Reply:
What I mean is:
```php
if ($currentPredicate instanceof CompositeExpression && $currentPredicate->getType() === $type) {
    return $currentPredicate->addMultiple($predicates);
}

if ($currentPredicate !== null) {
    array_unshift($predicates, $currentPredicate);
} elseif (count($predicates) === 1) {
    return $predicates[0];
}

return new CompositeExpression($type, $predicates);
```

Reply:
Indeed! Done in 3264682c463a87e17c76a5f0ec5ef13d3604ea2a.